### PR TITLE
[Serializer] forward exceptions caught in the `AbstractObjectNormalizer`

### LINKED ISSUE #57427: [Serializer] Upgrade to 7.1 - NotNormalizableValueException with $useMessageForUser=true - message lost
### Symfony version(s) affected

7.1.1

### Description

This concerns the denormalisation of a nested object using a custom (de)normalizer and `COLLECT_DENORMALIZATION_ERRORS=true`.

**In Symfony 6.4 & 7.0**, when a custom (de)normalizer throws a `NotNormalizableValueException` with `$useMessageForUser = true` in the `denormalize()` function, if Symfony Serializer’s context is set with `DenormalizerInterface::COLLECT_DENORMALIZATION_ERRORS => true`, it throws a `PartialDenormalizationException` and uses `message`, `path`, and `expectedTypes` defined in `NotNormalizableValueException::createForUnexpectedDataType()` to inform the user.

In **Symfony 7.1**, a broader message is generated using the class of the nested object, and loses the valuable information given by the custom normalizer.

----------

Code in a minimalistic Normalizer

```php
    public function denormalize(mixed $data, string $type, string $format = null, array $context = []): MySpecialDTO
    {
        $mySpecialDTO = new MySpecialDTO();

        $valueThatNeedsProcessing = $data['key'];
        if ('goodValue' !== $valueThatNeedsProcessing) {
            // ****************************************************************
            // The values & message set here used to be forwarded to client.
            // Not anymore on 7.1.
            // ****************************************************************
            throw NotNormalizableValueException::createForUnexpectedDataType('A clear message for the end user', 'Sample', ['ValidTypeIExpect'], $context['deserialization_path'].'.key', true);
        }

        $mySpecialDTO->key = $valueThatNeedsProcessing;

        return $mySpecialDTO;
    }
``` 

----------

**Symfony 6.4 & 7.0 output, with desired message :**

```json
{
    "type": "https:\/\/symfony.com\/errors\/validation",
    "title": "Validation Failed",
    "status": 422,
    "detail": "mySpecialField.key: This value should be of type ValidTypeIExpect.",
    "violations": [
        {
            "propertyPath": "mySpecialField.key",
            "title": "This value should be of type ValidTypeIExpect.",
            "template": "This value should be of type {{ type }}.",
            "parameters": {
                "{{ type }}": "ValidTypeIExpect"
            },
            "hint": "A clear message for the end user"
        }
    ]
}
```

**Symfony 7.1 output, with a broader message:**

```json
{
    "type": "https:\/\/symfony.com\/errors\/validation",
    "title": "Validation Failed",
    "status": 422,
    "detail": "mySpecialField: This value should be of type App\\DTO\\MySpecialDTO|null.",
    "violations": [
        {
            "propertyPath": "mySpecialField",
            "title": "This value should be of type App\\DTO\\MySpecialDTO|null.",
            "template": "This value should be of type {{ type }}.",
            "parameters": {
                "{{ type }}": "App\\DTO\\MySpecialDTO|null"
            }
        }
    ]
}
```

### How to reproduce

I wrote a simple reproducer, here : https://github.com/yblatti/repro-denorm-error

You'll find relevant code in `src/Controller`,  `src/DTO` and `src/Serialization`.

The same code is used, only the Symfony version changes accros 3 branches : [sf_6.4](https://github.com/yblatti/repro-denorm-error/tree/sf_6.4), [sf_7.0](https://github.com/yblatti/repro-denorm-error/tree/sf_7.0), [sf_7.1](https://github.com/yblatti/repro-denorm-error/tree/sf_7.1), 

**To test on different versions :** 

```shell
git clone https://github.com/yblatti/repro-denorm-error.git
cd repro-denorm-error

# Symfony 6.4
git checkout origin/sf_6.4
composer install
bin/console cache:clear --env=test
bin/phpunit --testdox
# all green results

# Symfony 7.0
git checkout origin/sf_7.0
composer install
bin/console cache:clear --env=test
bin/phpunit --testdox
# all green results

# Symfony 7.1
git checkout origin/sf_7.1
composer install
bin/console cache:clear --env=test
bin/phpunit --testdox
# the second test fails, as the message set in NotNormalizableValueException::createForUnexpectedDataType() is not used
```

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #57427
| License       | MIT

### CODE REVIEW

**Comment by xabbuh** [comment]
this makes sure that `validateAndDenormalize()` behaves the same way as before the introduction of the TypeInfo component (see the legacy implementation in `validateAndDenormalizeLegacy()`

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 1]
```diff
@@ -665,6 +668,18 @@ private function validateAndDenormalize(Type $type, string $currentClass, string
...
+
+                        if (TypeIdentifier::STRING === $typeIdentifier) {
+                            return '';
+                        }
+
+                        $isNullable = $isNullable ?: $type->isNullable();
+                    }
+
                     switch ($typeIdentifier) {
                         case TypeIdentifier::ARRAY:
```
Comment (HypeMC) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Isn't this case now redundant? I don't see it in the legacy implementation either.

Comment (xabbuh) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: indeed, good catch 👍

**Review by nicolas-grekas — COMMENTED** [review]
A small test case? :pray:

**Comment by HypeMC** [comment]
@nicolas-grekas @xabbuh See #57559, the test passes on 5.4, 6.4 and 7.0, but fails on 7.1. When I apply the changes made in this PR, it passes on 7.1 as well.

**Comment by xabbuh** [comment]
@HypeMC 👍 great, I have cherry-picked your commit here too

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**